add shadow-muting chat hiding feature (automatic flagging of public posts from shadow-muted players) - fixes 10851

Alys opened the pull request:
fixes https://github.com/HabitRPG/habitica/issues/10851

This PR allows moderators to mark a player as being shadow-muted, which will automatically hide public posts from that player. 

In detail:

- Mods can use the Hall to turn shadow-muting on and off:
![image](https://user-images.githubusercontent.com/1495809/59975550-6d538c80-95fc-11e9-948e-bf16e353f2e7.png)

- Mods can use icons in a player's profile to turn shadow-muting on and off. This PR also converts the profile admin tools from icons to links, and allows the crown icon to be used as a toggle to show or hide the links. These changes are only visible to moderators and staff. Regular users will not see anything different in the profile so normal visual design considerations aren't needed for this.

   - admin tools toggled off:
![image](https://user-images.githubusercontent.com/1495809/59975558-983de080-95fc-11e9-90d9-3bf993314873.png)
   - admin tools toggled on for a player who has not been muted or banned:
![image](https://user-images.githubusercontent.com/1495809/59975566-aab81a00-95fc-11e9-9941-b3321a34beae.png)
   - admin tools toggled on for a player who has been banned and had both types of muting applied (this also shows the on-hover effect for one of the links):
![image](https://user-images.githubusercontent.com/1495809/59975574-d0ddba00-95fc-11e9-88eb-91429410aed0.png)


When a player has been shadow-muted, the only function that's affected is posting to the Tavern and public guilds. All other functions in Habitica work normally. Any public or private posts made before the player was shadow-muted are not affected.

When a shadow-muted player posts to the Tavern or a public guild:
- Their post is automatically flagged to be hidden immediately.
- Only mods/staff and the player themself can see the post.
- The player sees no indication that their post is hidden.
- Mods see a message on the post (screenshot below).
- Mods can un-hide the post after they've reviewed it (in the same way they can un-hide manually-flagged posts).
- Mods receive a notification in Slack that a shadow-muted post has been made.
- Mods receive an email for the shadow-muted post. _(I still need to create the Mandrill template for this - I'll do that after/if this PR is approved.)_
![image](https://user-images.githubusercontent.com/1495809/59975624-798c1980-95fd-11e9-8be2-935b8f47df17.png)


The first three commits modify existing tests so you might want to review them separately:
1. 4127ac5b36bcece03335075879faca2b51601cb5 - moves two existing tests into the 'mute user' describe block, where there's already one test that they are related to
2. 9e0b663706356d9c71c13c06ca88204e41043b2d - changes those two existing tests to have names consistent with the other one in the mute block and to use const instead of let
3. a6f0ded2bc3923b8fff0c1c9c2c0536abca0f4f8 - improves methods for restoring chat permissions to test users (handles an old TODO comment)

The fourth commit (acdd3d88ad0f42c55ebeb9bf0a92ac53d379a9ca) adds new tests for shadow-muting. It also defines three new constants, all related to flagging / hiding messages:
```
export const CHAT_FLAG_LIMIT_FOR_HIDING = 2; // hide posts that have this many flags
export const CHAT_FLAG_FROM_MOD = 5; // a flag from a moderator counts as this many flags
export const CHAT_FLAG_FROM_SHADOW_MUTE = 10; // a shadow-muted user's post starts with this many flags
// @TODO use those constants to replace hard-coded numbers
```
I've used those constants in the code that I added and modified in this PR. but there's still other pieces of code that don't use them yet (code that's not directly related to this change). I'll adjust the remaining code to use the constants everywhere after/if this PR is accepted.

The remaining commits implement the shadow-muting feature itself.

When/if you're happy with the code, can it please be merged to delta so that I can test the email and slack notification code, and let the other mods test it?

Things that I need to do after/if this PR is accepted:
- [x] Create a Mandrill template `shadow-muted-post-report-to-mods`, similar to the `slur-report-to-mods template`.
- [x] Test the Slack message from delta or staging. It must clearly indicate that the post has been hidden automatically.
- [ ] Ask the other mods that they find the Slack notification to be clear enough.
- [x] Test the notification email from delta or staging. It must indicate that the post has been hidden automatically.
- [ ] Create an issue to remind myself to replace hard-coded numbers with the new constants (CHAT_FLAG_LIMIT_FOR_HIDING, CHAT_FLAG_FROM_MOD, CHAT_FLAG_FROM_SHADOW_MUTE)

---

_For anyone curious about how this shadow muting feature will be used, moderators are expecting to use it in only very rare, extreme cases. For example, if we're trying to control a persistent troll who is making new accounts to avoid a normal mute, we would shadow-mute them to decrease their impact on the community until staff have had time to set up an IP address ban._

_The primary reason we want this feature is because its code will become the basis of the feature described in https://github.com/HabitRPG/habitica/issues/8063_

_The current `flags.chatRevoked` mute feature will continue to be used in almost all cases where muting is appropriate._

Alys committed:
move existing tests for chatRevoked users to 'mute user' describe block

Alys committed:
give consistent names to chatRevoked tests and use const not let

Alys committed:
improve methods for restoring chat permissions to test users

Alys committed:
add tests for shadow-muting and define constants for flag-related numbers

Alys committed:
update user profile URLs and reverse private/public 'if' statements

Alys committed:
implement shadow muting in the API and schemas

Alys committed:
add interface for mods to turn shadow muting on/off for a user

- checkbox in the Hall
- icon in the user's profile

Alys committed:
mark chat posts as being shadow muted (marking is visible to mods only)

Alys committed:
convert Admin Tools in profile from icons to text; make crown icon a toggle

paglias commented on website/client/components/chat/chatCard.vue:
```diff
@@ -1,8 +1,9 @@
   .mentioned-icon(v-if='isUserMentioned')
-  .message-hidden(v-if='!inbox && msg.flagCount === 1 && user.contributor.admin') Message flagged once, not hidden
-  .message-hidden(v-if='!inbox && msg.flagCount > 1 && user.contributor.admin') Message hidden
+  .message-hidden(v-if='!inbox && user.contributor.admin && msg.flagCount === 1') Message flagged once, not hidden
```
these three `message-hidden ` might become a bit more readable if we could use only one and then display the string based on a computer prop that returns the string maybe?

Alys replied:
I think I've done that how you wanted it but if it's not exactly what you had in mind, tell me and I'll change it. https://github.com/HabitRPG/habitica/pull/11239/commits/180ae50fb74444a70a8e0ff639890ee280c5e182
You were right - it's nicer now!

The new code has a check for msg.flagCount being non-zero in two places (in the `.message-hidden(v-if` line and in `flagCountDescription()`) which is redundant but I think it's clearer this way. However I'm happy to remove the check from either location if you want.

I still need to create the mandrill template. I'll post again when that's done. Feel free to ignore this until then if you want. :)

paglias commented on website/server/controllers/api-v3/chat.js:
```diff
@@ -186,7 +187,43 @@ api.postChat = {
+        {name: 'GROUP_URL', content: groupUrl},
+      ];
+
+      sendTxn(FLAG_REPORT_EMAILS, 'shadow-muted-post-report-to-mods', report);
```
I'll have to check but we might have to add the template before the PR goes live or it might throw an error on the email server

paglias commented:
@Alys you want this go on delta for some this?

paglias commented:
Actually nevermind, I see that you asked for this to be put on delta. The code looks good but I think we should first add the template to mandrill

Alys commented:
Sure, no worries! I just didn't want to create the mandrill template before the PR had been reviewed in case any of the parameters needed to be changed. :) I'll post here when I've created it and made the code improvement you suggested.

Alys committed:
move logic for displaying flag count to a computed property

Alys committed:
Merge branch 'develop' into shadow-muting

paglias reviewed (commented):
lgtm!

Alys commented:
I've created the Mandrill template `shadow-muted-post-report-to-mods`.

This is ready to go to delta now, but there's no need to rush - just when you have time and it's convenient.

paglias commented:
pushed to delta!

Alys committed:
prevent chat notifications for shadow-muted posts

Alys committed:
Merge branch 'shadow-muting' of https://github.com/Alys/habitica into shadow-muting

Alys commented:
The slack notification and email is working well from delta (paglias, thanks for deploying!)

The other mods are testing now.

I've realised that when a post is automatically hidden because of shadow-muting, the new message notification should not be sent to all guild members because there's nothing new they can see. I've added a commit for that. In the test file, I've added a new `context` block around existing tests so you'll probably want to review it with whitespace differences ignored:
https://github.com/HabitRPG/habitica/pull/11239/commits/64a92669087632104d786e2ecb967195e4b4e7e7

Alys commented:
This has worked well for me in delta and none of the other mods have made requests for changes so this is ready to be merged if you're happy with the code.

It doesn't need to stay on delta any longer, so if you need to recreate delta again before this is merged, you don't have to add this PR to it again.
